File: ios/ClaviKeyboard/ClaviKeyboardExtension/TextFixEngine.swift
import Foundation

/**
 * Text fix engine — iOS mirror of TextFixEngine.kt (v0.4).
 *
 * Principle: FIX, DON'T REWRITE.
 * - Correct obvious typos and common spelling errors
 * - Fix punctuation spacing
 * - NEVER change word choice, sentence structure, tone, or style
 * - If unsure → return nil
 */
struct TextFixEngine {

    struct Fix {
        let original: String   // full textBefore BEFORE fix
        let fixed: String      // full textBefore AFTER fix
        let description: String
    }

    static func analyze(_ textBefore: String) -> Fix? {
        guard !textBefore.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }

        if let fix = fixDoubleSpace(textBefore)         { return fix }
        if let fix = fixSpaceBeforePunct(textBefore)    { return fix }
        if let fix = fixMissingSpaceAfterPunct(textBefore) { return fix }
        if let fix = fixUkrainianTypo(textBefore)       { return fix }
        if let fix = fixEnglishTypo(textBefore)         { return fix }
        if let fix = fixRepeatedChars(textBefore)       { return fix }
        if let fix = fixAccidentalCaps(textBefore)      { return fix }

        return nil
    }

    // MARK: - Fix implementations

    private static func fixDoubleSpace(_ text: String) -> Fix? {
        guard text.hasSuffix("  ") else { return nil }
        let fixed = text.replacingOccurrences(of: "  +$", with: " ", options: .regularExpression)
        return Fix(original: text, fixed: fixed, description: "double space")
    }

    private static func fixSpaceBeforePunct(_ text: String) -> Fix? {
        guard let range = text.range(of: #"\w( )([.,!?;:])$"#, options: .regularExpression) else { return nil }
        // Find the space group
        guard let spaceRange = text.range(of: #"(?<=\w)( )(?=[.,!?;:]$)"#, options: .regularExpression) else { return nil }
        var fixed = text
        fixed.removeSubrange(spaceRange)
        return Fix(original: text, fixed: fixed, description: "space before punctuation")
    }

    private static func fixMissingSpaceAfterPunct(_ text: String) -> Fix? {
        let pattern = #"[a-zA-Zа-яА-ЯіІїЇєЄґҐ]([.!?])([a-zA-Zа-яА-ЯіІїЇєЄґҐ])$"#
        guard let match = text.range(of: pattern, options: .regularExpression) else { return nil }
        // Find the second letter group and insert space before it
        guard let letterRange = text.range(of: #"(?<=[.!?])([a-zA-Zа-яА-ЯіІїЇєЄґҐ])$"#, options: .regularExpression) else { return nil }
        var fixed = text
        fixed.insert(" ", at: letterRange.lowerBound)
        return Fix(original: text, fixed: fixed, description: "missing space after sentence")
    }

    private static func fixUkrainianTypo(_ text: String) -> Fix? {
        guard let lastWord = extractLastWord(text) else { return nil }
        guard let corrected = ukTypos[lastWord.lowercased()] else { return nil }
        let prefix = String(text.dropLast(lastWord.count))
        let replacement = lastWord.first?.isUppercase == true
            ? corrected.prefix(1).uppercased() + corrected.dropFirst()
            : corrected
        return Fix(original: text, fixed: prefix + replacement, description: "typo: \(lastWord) → \(corrected)")
    }

    private static func fixEnglishTypo(_ text: String) -> Fix? {
        guard let lastWord = extractLastWord(text) else { return nil }
        guard let corrected = enTypos[lastWord.lowercased()] else { return nil }
        let prefix = String(text.dropLast(lastWord.count))
        let replacement = lastWord.first?.isUppercase == true
            ? corrected.prefix(1).uppercased() + corrected.dropFirst()
            : corrected
        return Fix(original: text, fixed: prefix + replacement, description: "typo: \(lastWord) → \(corrected)")
    }

    private static func fixRepeatedChars(_ text: String) -> Fix? {
        guard let lastWord = extractLastWord(text), lastWord.count >= 4 else { return nil }
        // Replace 3+ repeated chars with 1 (or 2 for Ukrainian doubled consonants)
        let pattern = #"(.)\1{2,}"#
        guard let fixed = try? NSRegularExpression(pattern: pattern).stringByReplacingMatches(
            in: lastWord,
            range: NSRange(lastWord.startIndex..., in: lastWord),
            withTemplate: { (match: String) -> String in
                let ch = String(match.first!)
                let ukDoubled = ["н","л","т","с","з"]
                return ukDoubled.contains(ch) ? ch + ch : ch
            }("")
        ) else { return nil }
        // NSRegularExpression doesn't support closure replacement — use manual approach
        return fixRepeatedCharsManual(text, lastWord: lastWord)
    }

    private static func fixRepeatedCharsManual(_ text: String, lastWord: String) -> Fix? {
        var fixed = ""
        var prev: Character? = nil
        var repeatCount = 0

        for ch in lastWord {
            if ch == prev {
                repeatCount += 1
                let ukDoubled: Set<Character> = ["н","л","т","с","з"]
                if repeatCount == 2 && ukDoubled.contains(ch) {
                    fixed.append(ch)  // allow one double
                } else if repeatCount >= 3 {
                    // skip — too many repeats
                } else {
                    fixed.append(ch)
                }
            } else {
                fixed.append(ch)
                prev = ch
                repeatCount = 1
            }
        }

        guard fixed != lastWord else { return nil }
        let prefix = String(text.dropLast(lastWord.count))
        return Fix(original: text, fixed: prefix + fixed, description: "repeated characters")
    }

    private static func fixAccidentalCaps(_ text: String) -> Fix? {
        guard let word = extractLastWord(text), word.count >= 4 else { return nil }
        let chars = Array(word)
        // Pattern: first two chars uppercase, at least one lowercase after (e.g. "ПРивіт", "HEllo")
        guard chars[0].isUppercase && chars[1].isUppercase else { return nil }
        guard chars.dropFirst(2).contains(where: { $0.isLowercase }) else { return nil }
        // Fix: keep first char uppercase, lowercase the rest
        let fixed = String(chars[0]) + String(word.dropFirst()).lowercased()
        guard fixed != word else { return nil }
        let prefix = String(text.dropLast(word.count))
        let shortWord = word.count > 6 ? String(word.prefix(6)) + "…" : word
        let shortFixed = fixed.count > 6 ? String(fixed.prefix(6)) + "…" : fixed
        return Fix(original: text, fixed: prefix + fixed,
                   description: "caps: \(shortWord) → \(shortFixed)")
    }

    private static func extractLastWord(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        let pattern = #"[\p{L}\p{M}'\-]+$"#
        guard let range = trimmed.range(of: pattern, options: .regularExpression) else { return nil }
        let word = String(trimmed[range])
        return word.count >= 2 ? word : nil
    }

    // MARK: - Typo dictionaries

    private static let ukTypos: [String: String] = [
        "pryvit":    "привіт",
        "dyakuyu":   "дякую",
        "будте":     "будьте",
        "памятати":  "пам'ятати",
        "компютер":  "комп'ютер",
        "звязок":    "зв'язок",
        "памятка":   "пам'ятка",
        "тчто":      "що",
        "зто":       "це",
    ]

    private static let enTypos: [String: String] = [
        "teh":         "the",
        "hte":         "the",
        "adn":         "and",
        "nad":         "and",
        "taht":        "that",
        "thta":        "that",
        "waht":        "what",
        "recieve":     "receive",
        "recieved":    "received",
        "beleive":     "believe",
        "freind":      "friend",
        "wierd":       "weird",
        "definately":  "definitely",
        "seperate":    "separate",
        "occured":     "occurred",
        "untill":      "until",
        "accomodate":  "accommodate",
        "occurence":   "occurrence",
        "tommorow":    "tomorrow",
        "tommorrow":   "tomorrow",
    ]
}

File: android/app/src/main/java/com/clavi/keyboard/TextFixEngine.kt
package com.clavi.keyboard

/**
 * Text fix engine — v0.4.
 *
 * Principle: FIX, DON'T REWRITE.
 * - Correct obvious typos and common spelling errors
 * - Fix punctuation spacing
 * - Fix capitalization after sentence end
 * - NEVER change word choice, sentence structure, tone, or style
 * - If unsure → don't fix (return null)
 *
 * Runs fully locally, no network, no cloud, no accounts.
 * Designed to be fast (<2ms) and minimal memory footprint.
 */
object TextFixEngine {

    data class Fix(
        val original: String,
        val fixed: String,
        val description: String,  // short human-readable reason
    )

    /**
     * Analyze the last word/phrase and return a fix if one is obvious.
     * Returns null if no fix is needed or confidence is too low.
     *
     * @param textBefore  text in the field before cursor
     */
    fun analyze(textBefore: String): Fix? {
        if (textBefore.isBlank()) return null

        // Fix 1: double space
        fixDoubleSpace(textBefore)?.let { return it }

        // Fix 2: space before punctuation
        fixSpaceBeforePunct(textBefore)?.let { return it }

        // Fix 3: missing space after sentence-ending punctuation
        fixMissingSpaceAfterPunct(textBefore)?.let { return it }

        // Fix 4: common Ukrainian typos (translit artifacts)
        fixUkrainianTypo(textBefore)?.let { return it }

        // Fix 5: common English typos
        fixEnglishTypo(textBefore)?.let { return it }

        // Fix 6: repeated characters (accidental double-tap)
        fixRepeatedChars(textBefore)?.let { return it }

        // Fix 7: accidental caps (e.g. "ПРивіт" → "Привіт")
        fixAccidentalCaps(textBefore)?.let { return it }

        return null
    }

    // ── Fix implementations ──

    private fun fixDoubleSpace(text: String): Fix? {
        if (!text.endsWith("  ")) return null
        return Fix(
            original = text,
            fixed = text.trimEnd(' ') + " ",
            description = "double space"
        )
    }

    private fun fixSpaceBeforePunct(text: String): Fix? {
        // Matches: word [space] [.,!?;:]
        val match = Regex("""\w( )([.,!?;:])$""").find(text) ?: return null
        val fixed = text.removeRange(match.groups[1]!!.range)
        return Fix(original = text, fixed = fixed, description = "space before punctuation")
    }

    private fun fixMissingSpaceAfterPunct(text: String): Fix? {
        // e.g. "hello.world" → "hello. world" — only if followed immediately by a letter
        val match = Regex("""[a-zA-Zа-яА-ЯіІїЇєЄґҐ]([.!?])([a-zA-Zа-яА-ЯіІїЇєЄґҐ])$""")
            .find(text) ?: return null
        val insertAt = match.groups[2]!!.range.first
        val fixed = text.substring(0, insertAt) + " " + text.substring(insertAt)
        return Fix(original = text, fixed = fixed, description = "missing space after sentence")
    }

    private fun fixUkrainianTypo(text: String): Fix? {
        val lastWord = extractLastWord(text) ?: return null
        val corrected = ukTypos[lastWord.lowercase()] ?: return null
        val fixedText = text.dropLast(lastWord.length) +
            if (lastWord[0].isUpperCase()) corrected.replaceFirstChar { it.uppercase() }
            else corrected
        return Fix(original = text, fixed = fixedText, description = "typo: $lastWord → $corrected")
    }

    private fun fixEnglishTypo(text: String): Fix? {
        val lastWord = extractLastWord(text) ?: return null
        val corrected = enTypos[lastWord.lowercase()] ?: return null
        val fixedText = text.dropLast(lastWord.length) +
            if (lastWord[0].isUpperCase()) corrected.replaceFirstChar { it.uppercase() }
            else corrected
        return Fix(original = text, fixed = fixedText, description = "typo: $lastWord → $corrected")
    }

    private fun fixRepeatedChars(text: String): Fix? {
        // e.g. "hellllo" → "hello", but NOT "ааа" (intentional emphasis) — only 3+ repeats
        val lastWord = extractLastWord(text) ?: return null
        if (lastWord.length < 4) return null
        val fixed = Regex("""(.)\1{2,}""").replace(lastWord) { m ->
            // Keep max 2 of the same char (some words have double like "ссилка")
            val char = m.groupValues[1]
            // Exception: don't fix Ukrainian doubled consonants нн, лл, тт etc.
            char.repeat(if (char.matches(Regex("[нлтссз]"))) 2 else 1)
        }
        if (fixed == lastWord) return null
        return Fix(
            original = text,
            fixed = text.dropLast(lastWord.length) + fixed,
            description = "repeated characters"
        )
    }

    private fun fixAccidentalCaps(text: String): Fix? {
        val word = extractLastWord(text) ?: return null
        if (word.length < 4) return null
        // Pattern: first two chars uppercase, at least one lowercase after (e.g. "ПРивіт", "HEllo")
        if (!word[0].isUpperCase() || !word[1].isUpperCase()) return null
        if (word.drop(2).none { it.isLowerCase() }) return null
        // Fix: keep first char uppercase, lowercase the rest
        val fixed = word[0].toString() + word.drop(1).lowercase()
        if (fixed == word) return null
        return Fix(
            original = text,
            fixed = text.dropLast(word.length) + fixed,
            description = "caps: ${word.take(6)}… → ${fixed.take(6)}…"
        )
    }

    private fun extractLastWord(text: String): String? {
        val trimmed = text.trimEnd()
        val match = Regex("""[\p{L}\p{M}'-]+$""").find(trimmed) ?: return null
        return match.value.takeIf { it.length >= 2 }
    }

    // ── Typo dictionaries ──
    // Small, hand-curated lists of the most common typos only.
    // NOT a full spell-checker — just the obvious high-confidence fixes.

    private val ukTypos: Map<String, String> = mapOf(
        // Transliteration artifacts (Latin → Cyrillic)
        "pryvit"    to "привіт",
        "dyakuyu"   to "дякую",
        "prosto"    to "просто",
        "dobre"     to "добре",
        "tak"       to "так",
        "shcho"     to "що",
        "yakshcho"  to "якщо",
        "tomu"      to "тому",
        "potim"     to "потім",
        "znayu"     to "знаю",
        "rozumiyu"  to "розумію",
        "bud"       to "буд",  // ambiguous — skip; keep for explicit "bud laska"
        // Missing apostrophe (most common Ukrainian typo category)
        "памятати"  to "пам'ятати",
        "памятка"   to "пам'ятка",
        "памяти"    to "пам'яті",
        "память"    to "пам'ять",
        "компютер"  to "комп'ютер",
        "звязок"    to "зв'язок",
        "звязати"   to "зв'язати",
        "мяч"       to "м'яч",
        "обєкт"     to "об'єкт",
        "субєкт"    to "суб'єкт",
        "девять"    to "дев'ять",
        "пять"      to "п'ять",
        "вязати"    to "в'язати",
        "вязаний"   to "в'язаний",
        "єднати"    to "з'єднати",
        "зєднати"   to "з'єднати",
        // Missing soft sign
        "будте"     to "будьте",
        "прийдіть"  to "прийдіть",  // valid — keep
        // Common keyboard slips (Cyrillic)
        "тчто"      to "що",
        "зто"       to "це",
        // Mixed Cyrillic/Latin
        "donт"      to "don't",
    )

    private val enTypos: Map<String, String> = mapOf(
        // Classic finger-roll transpositions
        "teh"         to "the",
        "hte"         to "the",
        "adn"         to "and",
        "nad"         to "and",
        "taht"        to "that",
        "thta"        to "that",
        "waht"        to "what",
        "hwat"        to "what",
        "yuo"         to "you",
        "youre"       to "you're",
        "dont"        to "don't",
        "doesnt"      to "doesn't",
        "didnt"       to "didn't",
        "wont"        to "won't",
        "cant"        to "can't",
        "isnt"        to "isn't",
        "wasnt"       to "wasn't",
        "werent"      to "weren't",
        "havent"      to "haven't",
        "hasnt"       to "hasn't",
        "wouldnt"     to "wouldn't",
        "couldnt"     to "couldn't",
        "shouldnt"    to "shouldn't",
        "im"          to "I'm",
        "ive"         to "I've",
        "id"          to "I'd",
        "ill"         to "I'll",
        // i-before-e violations
        "recieve"     to "receive",
        "recieved"    to "received",
        "beleive"     to "believe",
        "beleived"    to "believed",
        "freind"      to "friend",
        "wierd"       to "weird",
        "acheive"     to "achieve",
        "acheived"    to "achieved",
        "peice"       to "piece",
        "releive"     to "relieve",
        // Double-letter confusion
        "occured"     to "occurred",
        "occurence"   to "occurrence",
        "accomodate"  to "accommodate",
        "embarass"    to "embarrass",
        "harrass"     to "harass",
        "necesary"    to "necessary",
        "posession"   to "possession",
        "agressive"   to "aggressive",
        "profesional" to "professional",
        "recomend"    to "recommend",
        // -ately / -itely
        "definately"  to "definitely",
        "definetly"   to "definitely",
        "absolutley"  to "absolutely",
        "immediatley" to "immediately",
        // -ate / -ete confusion
        "seperate"    to "separate",
        "seperated"   to "separated",
        "desparate"   to "desperate",
        // Common misspellings
        "untill"      to "until",
        "tommorow"    to "tomorrow",
        "tommorrow"   to "tomorrow",
        "calender"    to "calendar",
        "cemetary"    to "cemetery",
        "goverment"   to "government",
        "medecine"    to "medicine",
        "millenium"   to "millennium",
        "miniscule"   to "minuscule",
        "mischevous"  to "mischievous",
        "neccessary"  to "necessary",
        "noticable"   to "noticeable",
        "priviledge"  to "privilege",
        "pronounciation" to "pronunciation",
        "publically"  to "publicly",
        "questionaire" to "questionnaire",
        "succesful"   to "successful",
        "suprise"     to "surprise",
        "tendancy"    to "tendency",
        "truely"      to "truly",
        "useable"     to "usable",
        "vaccum"      to "vacuum",
        "withold"     to "withhold",
        // Mixed Cyrillic/Latin keys (common on bilingual keyboards)
        "donт"        to "don't",
        "arт"         to "art",
        "isт"         to "ist",
    )
}
